<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Julia API · OnlinePCA.jl</title><meta name="title" content="Julia API · OnlinePCA.jl"/><meta property="og:title" content="Julia API · OnlinePCA.jl"/><meta property="twitter:title" content="Julia API · OnlinePCA.jl"/><meta name="description" content="Documentation for OnlinePCA.jl."/><meta property="og:description" content="Documentation for OnlinePCA.jl."/><meta property="twitter:description" content="Documentation for OnlinePCA.jl."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">OnlinePCA.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Julia API</a><ul class="internal"><li><a class="tocitem" href="#Binarization-(CSV-file)"><span>Binarization (CSV file)</span></a></li><li><a class="tocitem" href="#Binarization-(Matrix-Market-MM-file)"><span>Binarization (Matrix Market &lt;MM&gt; file)</span></a></li><li><a class="tocitem" href="#Binarization-(Binary-COO-BinCOO-file)"><span>Binarization (Binary COO &lt;BinCOO&gt; file)</span></a></li><li><a class="tocitem" href="#Summarization"><span>Summarization</span></a></li><li><a class="tocitem" href="#Filtering"><span>Filtering</span></a></li><li><a class="tocitem" href="#Identifying-Highly-Variable-Genes"><span>Identifying Highly Variable Genes</span></a></li><li><a class="tocitem" href="#GD-PCA"><span>GD-PCA</span></a></li><li><a class="tocitem" href="#SGD-PCA"><span>SGD-PCA</span></a></li><li><a class="tocitem" href="#Oja&#39;s-method"><span>Oja&#39;s method</span></a></li><li><a class="tocitem" href="#CCIPCA"><span>CCIPCA</span></a></li><li><a class="tocitem" href="#RSGD-PCA"><span>RSGD-PCA</span></a></li><li><a class="tocitem" href="#SVRG-PCA"><span>SVRG-PCA</span></a></li><li><a class="tocitem" href="#RSVRG-PCA"><span>RSVRG-PCA</span></a></li><li><a class="tocitem" href="#Orthogonal-Iteration-(Power-method)"><span>Orthogonal Iteration (Power method)</span></a></li><li><a class="tocitem" href="#Arnoldi-method"><span>Arnoldi method</span></a></li><li><a class="tocitem" href="#Lanczos-method"><span>Lanczos method</span></a></li><li><a class="tocitem" href="#Halko&#39;s-method"><span>Halko&#39;s method</span></a></li><li><a class="tocitem" href="#Algorithm-971"><span>Algorithm 971</span></a></li><li><a class="tocitem" href="#Randomized-Block-Krylov-Iteration"><span>Randomized Block Krylov Iteration</span></a></li><li><a class="tocitem" href="#Single-pass-PCA-type-I"><span>Single-pass PCA type I</span></a></li><li><a class="tocitem" href="#Single-pass-PCA-type-II"><span>Single-pass PCA type II</span></a></li><li><a class="tocitem" href="#Summarization-for-10X-HDF5"><span>Summarization for 10X-HDF5</span></a></li><li><a class="tocitem" href="#ALGORITHM971-for-10X-HDF5"><span>ALGORITHM971 for 10X-HDF5</span></a></li><li><a class="tocitem" href="#Sparse-Randomized-SVD-(ALGORITHM971-for-Binarized-MM-file)"><span>Sparse Randomized SVD (ALGORITHM971 for Binarized MM file)</span></a></li><li><a class="tocitem" href="#Exact-Out-of-Core-PCA"><span>Exact Out-of-Core PCA</span></a></li></ul></li><li><a class="tocitem" href="../commandline/">Command line Tool</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Julia API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Julia API</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/rikenbit/OnlinePCA.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/rikenbit/OnlinePCA.jl/blob/master/docs/src/juliaapi.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="OnlinePCA.jl-(Julia-API)"><a class="docs-heading-anchor" href="#OnlinePCA.jl-(Julia-API)">OnlinePCA.jl (Julia API)</a><a id="OnlinePCA.jl-(Julia-API)-1"></a><a class="docs-heading-anchor-permalink" href="#OnlinePCA.jl-(Julia-API)" title="Permalink"></a></h1><h2 id="Binarization-(CSV-file)"><a class="docs-heading-anchor" href="#Binarization-(CSV-file)">Binarization (CSV file)</a><a id="Binarization-(CSV-file)-1"></a><a class="docs-heading-anchor-permalink" href="#Binarization-(CSV-file)" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.csv2bin-Tuple{}" href="#OnlinePCA.csv2bin-Tuple{}"><code>OnlinePCA.csv2bin</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">csv2bin(;csvfile::AbstractString=&quot;&quot;, binfile::AbstractString=&quot;&quot;)</code></pre><p>Convert a CSV file to Julia Binary file.</p><p><code>csvfile</code> and <code>binfile</code> are specified such as Data.csv and Data.dat, respectively.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/csv2bin.jl#L1-L7">source</a></section></article><h2 id="Binarization-(Matrix-Market-MM-file)"><a class="docs-heading-anchor" href="#Binarization-(Matrix-Market-MM-file)">Binarization (Matrix Market &lt;MM&gt; file)</a><a id="Binarization-(Matrix-Market-MM-file)-1"></a><a class="docs-heading-anchor-permalink" href="#Binarization-(Matrix-Market-MM-file)" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.mm2bin-Tuple{}" href="#OnlinePCA.mm2bin-Tuple{}"><code>OnlinePCA.mm2bin</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>mm2bin(;mmfile::AbstractString=&quot;&quot;, binfile::AbstractString=&quot;&quot;)</p><p>Convert a Matrix Market (MM) file to Julia Binary file.</p><p><strong>Input Arguments</strong></p><ul><li><code>mmfile</code> : Matrix Market file (e.g., Data.mtx).</li><li><code>binfile</code> : Julia Binary file (e.g., Data.mtx.zst).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/mm2bin.jl#L1-L10">source</a></section></article><h2 id="Binarization-(Binary-COO-BinCOO-file)"><a class="docs-heading-anchor" href="#Binarization-(Binary-COO-BinCOO-file)">Binarization (Binary COO &lt;BinCOO&gt; file)</a><a id="Binarization-(Binary-COO-BinCOO-file)-1"></a><a class="docs-heading-anchor-permalink" href="#Binarization-(Binary-COO-BinCOO-file)" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.bincoo2bin-Tuple{}" href="#OnlinePCA.bincoo2bin-Tuple{}"><code>OnlinePCA.bincoo2bin</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>bincoo2bin(;bincoofile::AbstractString=&quot;&quot;, binfile::AbstractString=&quot;&quot;)</p><p>Convert a Binary COO (BinCOO) file to Julia Binary file.</p><p><strong>Input Arguments</strong></p><ul><li><code>bincoofile</code> : Binary COO file (e.g., Data.bincoo).</li><li><code>binfile</code> : Julia Binary file (e.g., Data.bincoo.zst).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/bincoo2bin.jl#L1-L10">source</a></section></article><h2 id="Summarization"><a class="docs-heading-anchor" href="#Summarization">Summarization</a><a id="Summarization-1"></a><a class="docs-heading-anchor-permalink" href="#Summarization" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.sumr-Tuple{}" href="#OnlinePCA.sumr-Tuple{}"><code>OnlinePCA.sumr</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">sumr(; binfile::AbstractString=&quot;&quot;, outdir::AbstractString=&quot;.&quot;, pseudocount::Number=1.0, mode::AbstractString=&quot;dense&quot;, chunksize::Int=1)</code></pre><p>Extract the summary information of data matrix.</p><p><strong>Input Arguments</strong></p><ul><li><code>binfile</code> is a Julia Binary file generated by <code>csv2bin</code> function.</li><li><code>outdir</code> is specified the directory you want to save the result.</li><li><code>pseudocount</code> is specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>mode</code> : &quot;dense&quot; or &quot;sparse_mm&quot; can be specified.</li><li><code>chunksize</code> : The number of rows to be read at once.</li></ul><p><strong>Output Files</strong></p><ul><li><code>Sample_NoCounts.csv</code> : Sum of counts in each column.</li><li><code>Feature_Means.csv</code> : Mean in each row.</li><li><code>Feature_LogMeans.csv</code> : Log10(Mean+pseudocount) in each row.</li><li><code>Feature_FTTMeans.csv</code> : FTT(Mean+pseudocount) in each row.</li><li><code>Feature_Vars.csv</code> : Sample variance in each row.</li><li><code>Feature_LogVars.csv</code> : Log10(Var+pseudocount) in each row.</li><li><code>Feature_FTTVars.csv</code> : FTT(Var+pseudocount) in each row.</li><li><code>Feature_CV2s.csv</code> : Coefficient of Variation in each row.</li><li><code>Feature_NoZeros.csv</code> : Number of zero-elements in each row.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/sumr.jl#L1-L25">source</a></section></article><h2 id="Filtering"><a class="docs-heading-anchor" href="#Filtering">Filtering</a><a id="Filtering-1"></a><a class="docs-heading-anchor-permalink" href="#Filtering" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.filtering-Tuple{}" href="#OnlinePCA.filtering-Tuple{}"><code>OnlinePCA.filtering</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filtering(;input::AbstractString=&quot;&quot;, featurelist::AbstractString=&quot;&quot;, samplelist::AbstractString=&quot;&quot;, thr1::Number=0.0, thr2::Number=0.0, direct1::AbstractString=&quot;+&quot;, direct2::AbstractString=&quot;+&quot;, outdir::AbstractString=&quot;.&quot;)</code></pre><p>This function filters the genes by some standards such as mean or variance of the genes.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : A Julia Binary file generated by <code>csv2bin</code> function.</li><li><code>featurelist</code> : A row-wise summary data such as. The CSV files are generated by <code>csv2bin</code> function.</li><li><code>thr</code> : The threshold to reject low-signal feature.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li></ul><p><strong>Output Files</strong></p><ul><li><code>filtered.zst</code> : Filtered binary file.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/filtering.jl#L1-L16">source</a></section></article><h2 id="Identifying-Highly-Variable-Genes"><a class="docs-heading-anchor" href="#Identifying-Highly-Variable-Genes">Identifying Highly Variable Genes</a><a id="Identifying-Highly-Variable-Genes-1"></a><a class="docs-heading-anchor-permalink" href="#Identifying-Highly-Variable-Genes" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.hvg-Tuple{}" href="#OnlinePCA.hvg-Tuple{}"><code>OnlinePCA.hvg</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">hvg(;binfile::AbstractString=&quot;&quot;, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, rowcv2list::AbstractString=&quot;&quot;, outdir::AbstractString=&quot;.&quot;)</code></pre><p>This function perform highly variable genes, which is an feature selection method in scRNA-seq studies.</p><p><strong>Input Arguments</strong></p><ul><li><code>binfile</code> is a Julia Binary file generated by <code>csv2bin</code> function.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>sumr</code> functions.</li><li><code>rowcv2list</code> : The cv2 of each row of matrix. The CSV file is generated by <code>sumr</code> functions.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li></ul><p><strong>Output Files</strong></p><ul><li><code>HVG_useForFit.csv</code> : Parameters to estimate the highly variable genes.</li><li><code>HVG_a0.csv</code> : Parameters to estimate the highly variable genes.</li><li><code>HVG_a1.csv</code> : Parameters to estimate the highly variable genes.</li><li><code>HVG_afits.csv</code> : Parameters to estimate the highly variable genes.</li><li><code>HVG_varFitRatio.csv</code> : Parameters to estimate the highly variable genes.</li><li><code>HVG_df.csv</code> : Parameters to estimate the highly variable genes.</li><li><code>HVG_pval.csv</code> : Parameters to estimate the highly variable genes.</li></ul><p><strong>Reference</strong></p><ul><li><a href="http://pklab.med.harvard.edu/scw2014/subpop_tutorial.html">Highly Variable Genes</a></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/hvg.jl#L1-L27">source</a></section></article><h2 id="GD-PCA"><a class="docs-heading-anchor" href="#GD-PCA">GD-PCA</a><a id="GD-PCA-1"></a><a class="docs-heading-anchor-permalink" href="#GD-PCA" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.gd-Tuple{}" href="#OnlinePCA.gd-Tuple{}"><code>OnlinePCA.gd</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">gd(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;,colsumlist::AbstractString=&quot;&quot;, dim::Number=3, stepsize::Number=0.1, numepoch::Number=3, scheduling::AbstractString=&quot;robbins-monro&quot;, g::Number=0.9, epsilon::Number=1.0e-8, lower::Number=0, upper::Number=1.0f+38, expvar::Number=0.1f0, evalfreq::Number=5000, offsetFull::Number=1f-20, offsetStoch::Number=1f-6, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Gradient descent method.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>stepsize</code> : The parameter used in every iteration.</li><li><code>numepoch</code> : The number of epoch.</li><li><code>scheduling</code> : Learning parameter scheduling. <code>robbins-monro</code>, <code>momentum</code>, <code>nag</code>, and <code>adagrad</code> are available.</li><li><code>g</code> : The parameter that is used when scheduling is specified as nag.</li><li><code>epsilon</code> : The parameter that is used when scheduling is specified as adagrad.</li><li><code>lower</code> : Stopping Criteria (When the relative change of error is below this value, the calculation is terminated)</li><li><code>upper</code> : Stopping Criteria (When the relative change of error is above this value, the calculation is terminated)</li><li><code>evalfreq</code> : Evaluation Frequency of Reconstruction Error</li><li><code>offsetFull</code> : Off set value for avoding overflow when calculating full gradient</li><li><code>offsetStoch</code> : Off set value for avoding overflow when calculating stochastic gradient</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every 1000 iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/gd.jl#L1-L42">source</a></section></article><h2 id="SGD-PCA"><a class="docs-heading-anchor" href="#SGD-PCA">SGD-PCA</a><a id="SGD-PCA-1"></a><a class="docs-heading-anchor-permalink" href="#SGD-PCA" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.sgd-Tuple{}" href="#OnlinePCA.sgd-Tuple{}"><code>OnlinePCA.sgd</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">sgd(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, colsumlist::AbstractString=&quot;&quot;, dim::Number=3, stepsize::Number=0.1, numbatch::Number=100, numepoch::Number=3, scheduling::AbstractString=&quot;robbins-monro&quot;, g::Number=0.9, epsilon::Number=1.0e-8, lower::Number=0, upper::Number=1.0f+38, expvar::Number=0.1f0, evalfreq::Number=5000, offsetFull::Number=1f-20, offsetStoch::Number=1f-6, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Stochastic gradient descent method.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>stepsize</code> : The parameter used in every iteration.</li><li><code>numbatch</code> : The number of batch size.</li><li><code>numepoch</code> : The number of epoch.</li><li><code>scheduling</code> : Learning parameter scheduling. <code>robbins-monro</code>, <code>momentum</code>, <code>nag</code>, and <code>adagrad</code> are available.</li><li><code>g</code> : The parameter that is used when scheduling is specified as nag.</li><li><code>epsilon</code> : The parameter that is used when scheduling is specified as adagrad.</li><li><code>lower</code> : Stopping Criteria (When the relative change of error is below this value, the calculation is terminated)</li><li><code>upper</code> : Stopping Criteria (When the relative change of error is above this value, the calculation is terminated)</li><li><code>evalfreq</code> : Evaluation Frequency of Reconstruction Error</li><li><code>offsetFull</code> : Off set value for avoding overflow when calculating full gradient</li><li><code>offsetStoch</code> : Off set value for avoding overflow when calculating stochastic gradient</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/sgd.jl#L1-L42">source</a></section></article><h2 id="Oja&#39;s-method"><a class="docs-heading-anchor" href="#Oja&#39;s-method">Oja&#39;s method</a><a id="Oja&#39;s-method-1"></a><a class="docs-heading-anchor-permalink" href="#Oja&#39;s-method" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.oja-Tuple{}" href="#OnlinePCA.oja-Tuple{}"><code>OnlinePCA.oja</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">oja(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, colsumlist::AbstractString=&quot;&quot;, dim::Number=3, stepsize::Number=0.1, numepoch::Number=3, scheduling::AbstractString=&quot;robbins-monro&quot;, g::Number=0.9, epsilon::Number=1.0e-8, lower::Number=0, upper::Number=1.0f+38, expvar::Number=0.1f0, evalfreq::Number=5000, offsetFull::Number=1f-20, offsetStoch::Number=1f-6, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Oja&#39;s method.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>stepsize</code> : The parameter used in every iteration.</li><li><code>numepoch</code> : The number of epoch.</li><li><code>scheduling</code> : Learning parameter scheduling. <code>robbins-monro</code>, <code>momentum</code>, <code>nag</code>, and <code>adagrad</code> are available.</li><li><code>g</code> : The parameter that is used when scheduling is specified as nag.</li><li><code>epsilon</code> : The parameter that is used when scheduling is specified as adagrad.</li><li><code>lower</code> : Stopping Criteria (When the relative change of error is below this value, the calculation is terminated)</li><li><code>upper</code> : Stopping Criteria (When the relative change of error is above this value, the calculation is terminated)</li><li><code>evalfreq</code> : Evaluation Frequency of Reconstruction Error</li><li><code>offsetFull</code> : Off set value for avoding overflow when calculating full gradient</li><li><code>offsetStoch</code> : Off set value for avoding overflow when calculating stochastic gradient</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul><p><strong>Reference</strong></p><ul><li>SGD-PCA（Oja&#39;s method) : <a href="https://www.sciencedirect.com/science/article/pii/0022247X85901313">Erkki Oja et. al., 1985</a>, <a href="https://www.sciencedirect.com/science/article/pii/S0893608005800899">Erkki Oja, 1992</a></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/oja.jl#L1-L45">source</a></section></article><h2 id="CCIPCA"><a class="docs-heading-anchor" href="#CCIPCA">CCIPCA</a><a id="CCIPCA-1"></a><a class="docs-heading-anchor-permalink" href="#CCIPCA" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.ccipca-Tuple{}" href="#OnlinePCA.ccipca-Tuple{}"><code>OnlinePCA.ccipca</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ccipca(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;,colsumlist::AbstractString=&quot;&quot;, dim::Number=3, stepsize::Number=0, numepoch::Number=3, lower::Number=0, upper::Number=1.0f+38, expvar::Number=0.1f0, evalfreq::Number=5000, offsetStoch::Number=1f-6, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Candid covariance-free incremental PCA.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>stepsize</code> : The parameter used in every iteration.</li><li><code>numepoch</code> : The number of epoch.</li><li><code>lower</code> : Stopping Criteria (When the relative change of error is below this value, the calculation is terminated)</li><li><code>upper</code> : Stopping Criteria (When the relative change of error is above this value, the calculation is terminated)</li><li><code>evalfreq</code> : Evaluation Frequency of Reconstruction Error</li><li><code>offsetStoch</code> : Off set value for avoding overflow when calculating stochastic gradient</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul><p><strong>Reference</strong></p><ul><li>CCIPCA : <a href="http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.7.5665&amp;rep=rep1&amp;type=pdf">Juyang Weng et. al., 2003</a></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/ccipca.jl#L1-L41">source</a></section></article><h2 id="RSGD-PCA"><a class="docs-heading-anchor" href="#RSGD-PCA">RSGD-PCA</a><a id="RSGD-PCA-1"></a><a class="docs-heading-anchor-permalink" href="#RSGD-PCA" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.rsgd-Tuple{}" href="#OnlinePCA.rsgd-Tuple{}"><code>OnlinePCA.rsgd</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">rsgd(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;,colsumlist::AbstractString=&quot;&quot;, dim::Number=3, stepsize::Number=0.1, numbatch::Number=100, numepoch::Number=3, scheduling::AbstractString=&quot;robbins-monro&quot;, g::Number=0.9, epsilon::Number=1.0e-8, lower::Number=0, upper::Number=1.0f+38, expvar::Number=0.1f0, evalfreq::Number=5000, offsetFull::Number=1f-20, offsetStoch::Number=1f-6, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Riemannian stochastic gradient descent method.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>stepsize</code> : The parameter used in every iteration.</li><li><code>numbatch</code> : The number of batch size.</li><li><code>numepoch</code> : The number of epoch.</li><li><code>scheduling</code> : Learning parameter scheduling. <code>robbins-monro</code>, <code>momentum</code>, <code>nag</code>, and <code>adagrad</code> are available.</li><li><code>g</code> : The parameter that is used when scheduling is specified as nag.</li><li><code>epsilon</code> : The parameter that is used when scheduling is specified as adagrad.</li><li><code>lower</code> : Stopping Criteria (When the relative change of error is below this value, the calculation is terminated)</li><li><code>upper</code> : Stopping Criteria (When the relative change of error is above this value, the calculation is terminated)</li><li><code>evalfreq</code> : Evaluation Frequency of Reconstruction Error</li><li><code>offsetFull</code> : Off set value for avoding overflow when calculating full gradient</li><li><code>offsetStoch</code> : Off set value for avoding overflow when calculating stochastic gradient</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul><p><strong>Reference</strong></p><ul><li>RSGD-PCA : <a href="https://arxiv.org/abs/1111.5280">Silvere Bonnabel, 2013</a></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/rsgd.jl#L1-L46">source</a></section></article><h2 id="SVRG-PCA"><a class="docs-heading-anchor" href="#SVRG-PCA">SVRG-PCA</a><a id="SVRG-PCA-1"></a><a class="docs-heading-anchor-permalink" href="#SVRG-PCA" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.svrg-Tuple{}" href="#OnlinePCA.svrg-Tuple{}"><code>OnlinePCA.svrg</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">svrg(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, colsumlist::AbstractString=&quot;&quot;, dim::Number=3, stepsize::Number=0.1, numbatch::Number=100, numepoch::Number=3, scheduling::AbstractString=&quot;robbins-monro&quot;, g::Number=0.9, epsilon::Number=1.0e-8, lower::Number=0, upper::Number=1.0f+38, expvar::Number=0.1f0, evalfreq::Number=5000, offsetFull::Number=1f-20, offsetStoch::Number=1f-6, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Variance-reduced stochastic gradient descent method, also known as VR-PCA.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>stepsize</code> : The parameter used in every iteration.</li><li><code>numbatch</code> : The number of batch size.</li><li><code>numepoch</code> : The number of epoch.</li><li><code>scheduling</code> : Learning parameter scheduling. <code>robbins-monro</code>, <code>momentum</code>, <code>nag</code>, and <code>adagrad</code> are available.</li><li><code>g</code> : The parameter that is used when scheduling is specified as nag.</li><li><code>epsilon</code> : The parameter that is used when scheduling is specified as adagrad.</li><li><code>lower</code> : Stopping Criteria (When the relative change of error is below this value, the calculation is terminated)</li><li><code>upper</code> : Stopping Criteria (When the relative change of error is above this value, the calculation is terminated)</li><li><code>evalfreq</code> : Evaluation Frequency of Reconstruction Error</li><li><code>offsetFull</code> : Off set value for avoding overflow when calculating full gradient</li><li><code>offsetStoch</code> : Off set value for avoding overflow when calculating stochastic gradient</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul><p><strong>Reference</strong></p><ul><li>SVRG-PCA : <a href="http://proceedings.mlr.press/v37/shamir15.pdf">Ohad Shamir, 2015</a></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/svrg.jl#L1-L46">source</a></section></article><h2 id="RSVRG-PCA"><a class="docs-heading-anchor" href="#RSVRG-PCA">RSVRG-PCA</a><a id="RSVRG-PCA-1"></a><a class="docs-heading-anchor-permalink" href="#RSVRG-PCA" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.rsvrg-Tuple{}" href="#OnlinePCA.rsvrg-Tuple{}"><code>OnlinePCA.rsvrg</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">rsvrg(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;,colsumlist::AbstractString=&quot;&quot;, dim::Number=3, stepsize::Number=0.1, numbatch::Number=100, numepoch::Number=3, scheduling::AbstractString=&quot;robbins-monro&quot;, g::Number=0.9, epsilon::Number=1.0e-8, lower::Number=0, upper::Number=1.0f+38, expvar::Number=0.1f0, evalfreq::Number=5000, offsetFull::Number=1f-20, offsetStoch::Number=1f-6, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Riemannian variance-reduced stochastic gradient descent method.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>stepsize</code> : The parameter used in every iteration.</li><li><code>numbatch</code> : The number of batch size.</li><li><code>numepoch</code> : The number of epoch.</li><li><code>scheduling</code> : Learning parameter scheduling. <code>robbins-monro</code>, <code>momentum</code>, <code>nag</code>, and <code>adagrad</code> are available.</li><li><code>g</code> : The parameter that is used when scheduling is specified as nag.</li><li><code>epsilon</code> : The parameter that is used when scheduling is specified as adagrad.</li><li><code>lower</code> : Stopping Criteria (When the relative change of error is below this value, the calculation is terminated)</li><li><code>upper</code> : Stopping Criteria (When the relative change of error is above this value, the calculation is terminated)</li><li><code>evalfreq</code> : Evaluation Frequency of Reconstruction Error</li><li><code>offsetFull</code> : Off set value for avoding overflow when calculating full gradient</li><li><code>offsetStoch</code> : Off set value for avoding overflow when calculating stochastic gradient</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul><p><strong>Reference</strong></p><ul><li>RSVRG-PCA : <a href="http://papers.nips.cc/paper/6515-riemannian-svrg-fast-stochastic-optimization-on-riemannian-manifolds.pdf">Hongyi Zhang, et. al., 2016</a>, <a href="https://arxiv.org/abs/1702.05594">Hiroyuki Sato, et. al., 2017</a></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/rsvrg.jl#L1-L46">source</a></section></article><h2 id="Orthogonal-Iteration-(Power-method)"><a class="docs-heading-anchor" href="#Orthogonal-Iteration-(Power-method)">Orthogonal Iteration (Power method)</a><a id="Orthogonal-Iteration-(Power-method)-1"></a><a class="docs-heading-anchor-permalink" href="#Orthogonal-Iteration-(Power-method)" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.orthiter-Tuple{}" href="#OnlinePCA.orthiter-Tuple{}"><code>OnlinePCA.orthiter</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">orthiter(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;,colsumlist::AbstractString=&quot;&quot;, dim::Number=3, numepoch::Number=10, lower::Number=0, upper::Number=1.0f+38, expvar::Number=0.1f0, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Orthogonal iteration, also known as block power method, subspace iteration or simultaneous iteration, ...etc.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>numepoch</code> : The number of epoch.</li><li><code>lower</code> : Stopping Criteria (When the relative change of error is below this value, the calculation is terminated)</li><li><code>upper</code> : Stopping Criteria (When the relative change of error is above this value, the calculation is terminated)</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/orthiter.jl#L1-L34">source</a></section></article><h2 id="Arnoldi-method"><a class="docs-heading-anchor" href="#Arnoldi-method">Arnoldi method</a><a id="Arnoldi-method-1"></a><a class="docs-heading-anchor-permalink" href="#Arnoldi-method" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.arnoldi-Tuple{}" href="#OnlinePCA.arnoldi-Tuple{}"><code>OnlinePCA.arnoldi</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">arnoldi(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;,colsumlist::AbstractString=&quot;&quot;, dim::Number=3, numepoch::Number=10, perm::Bool=false, cper::Number=1f0)</code></pre><p>Arnoldi method.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/arnoldi.jl#L1-L28">source</a></section></article><h2 id="Lanczos-method"><a class="docs-heading-anchor" href="#Lanczos-method">Lanczos method</a><a id="Lanczos-method-1"></a><a class="docs-heading-anchor-permalink" href="#Lanczos-method" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.lanczos-Tuple{}" href="#OnlinePCA.lanczos-Tuple{}"><code>OnlinePCA.lanczos</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">lanczos(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;,colsumlist::AbstractString=&quot;&quot;, dim::Number=3, numepoch::Number=10, perm::Bool=false, cper::Number=1f0)</code></pre><p>Lanczos method.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/lanczos.jl#L1-L28">source</a></section></article><h2 id="Halko&#39;s-method"><a class="docs-heading-anchor" href="#Halko&#39;s-method">Halko&#39;s method</a><a id="Halko&#39;s-method-1"></a><a class="docs-heading-anchor-permalink" href="#Halko&#39;s-method" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.halko-Tuple{}" href="#OnlinePCA.halko-Tuple{}"><code>OnlinePCA.halko</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">halko(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, colsumlist::AbstractString=&quot;&quot;, dim::Number=3, noversamples::Number=5, niter::Number=3, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Halko&#39;s method, which is one of randomized SVD algorithm.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>noversamples</code> : The number of over-sampling.</li><li><code>niter</code> : The number of power interation.</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>V</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>U</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/halko.jl#L1-L32">source</a></section></article><h2 id="Algorithm-971"><a class="docs-heading-anchor" href="#Algorithm-971">Algorithm 971</a><a id="Algorithm-971-1"></a><a class="docs-heading-anchor-permalink" href="#Algorithm-971" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.algorithm971-Tuple{}" href="#OnlinePCA.algorithm971-Tuple{}"><code>OnlinePCA.algorithm971</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">algorithm971(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, colsumlist::AbstractString=&quot;&quot;, dim::Number=3, noversamples::Number=5, niter::Number=3, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Algorithm 971, which is one of randomized SVD algorithm.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>noversamples</code> : The number of over-sampling.</li><li><code>niter</code> : The number of power interation.</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>V</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>U</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/algorithm971.jl#L1-L32">source</a></section></article><h2 id="Randomized-Block-Krylov-Iteration"><a class="docs-heading-anchor" href="#Randomized-Block-Krylov-Iteration">Randomized Block Krylov Iteration</a><a id="Randomized-Block-Krylov-Iteration-1"></a><a class="docs-heading-anchor-permalink" href="#Randomized-Block-Krylov-Iteration" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.rbkiter-Tuple{}" href="#OnlinePCA.rbkiter-Tuple{}"><code>OnlinePCA.rbkiter</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">rbkiter(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;,colsumlist::AbstractString=&quot;&quot;, dim::Number=3, numepoch::Number=10, lower::Number=0, upper::Number=1.0f+38, expvar::Number=0.1f0, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Randomized Block Krylov Iteration.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>numepoch</code> : The number of epoch.</li><li><code>lower</code> : Stopping Criteria (When the relative change of error is below this value, the calculation is terminated)</li><li><code>upper</code> : Stopping Criteria (When the relative change of error is above this value, the calculation is terminated)</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>W</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>V</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li><li>stop : Whether the calculation is converged</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/rbkiter.jl#L1-L35">source</a></section></article><h2 id="Single-pass-PCA-type-I"><a class="docs-heading-anchor" href="#Single-pass-PCA-type-I">Single-pass PCA type I</a><a id="Single-pass-PCA-type-I-1"></a><a class="docs-heading-anchor-permalink" href="#Single-pass-PCA-type-I" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.singlepass-Tuple{}" href="#OnlinePCA.singlepass-Tuple{}"><code>OnlinePCA.singlepass</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">singlepass(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, colsumlist::AbstractString=&quot;&quot;, dim::Number=3, noversamples::Number=5, niter::Number=3, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Single-pass PCA type I, which is one of randomized SVD algorithm.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>noversamples</code> : The number of over-sampling.</li><li><code>niter</code> : The number of power interation.</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>V</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>U</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/singlepass.jl#L1-L32">source</a></section></article><h2 id="Single-pass-PCA-type-II"><a class="docs-heading-anchor" href="#Single-pass-PCA-type-II">Single-pass PCA type II</a><a id="Single-pass-PCA-type-II-1"></a><a class="docs-heading-anchor-permalink" href="#Single-pass-PCA-type-II" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.singlepass2-Tuple{}" href="#OnlinePCA.singlepass2-Tuple{}"><code>OnlinePCA.singlepass2</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">singlepass2(;input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, pseudocount::Number=1f0, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, colsumlist::AbstractString=&quot;&quot;, dim::Number=3, noversamples::Number=5, niter::Number=3, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1f0)</code></pre><p>Single-pass PCA type II, which is one of randomized SVD algorithm.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {log,ftt,raw}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>noversamples</code> : The number of over-sampling.</li><li><code>niter</code> : The number of power interation.</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>V</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>U</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/singlepass2.jl#L1-L32">source</a></section></article><h2 id="Summarization-for-10X-HDF5"><a class="docs-heading-anchor" href="#Summarization-for-10X-HDF5">Summarization for 10X-HDF5</a><a id="Summarization-for-10X-HDF5-1"></a><a class="docs-heading-anchor-permalink" href="#Summarization-for-10X-HDF5" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.tenxsumr-Tuple{}" href="#OnlinePCA.tenxsumr-Tuple{}"><code>OnlinePCA.tenxsumr</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">tenxsumr(; tenxfile::AbstractString=&quot;&quot;, outdir::AbstractString=&quot;.&quot;, group::AbstractString=&quot;&quot;, chunksize::Number=5000)</code></pre><p>Extract the summary information of 10X-HDF5.</p><p><strong>Input Arguments</strong></p><ul><li><code>tenxfile</code> is the HDF5 file formatted by 10X Genomics.</li><li><code>outdir</code> is specified the directory you want to save the result.</li><li><code>group</code> is the group name of HDF5 (e.g. mm10).</li><li><code>chunksize</code> is the number of rows reading at once (e.g. 5000).</li></ul><p><strong>Output Files</strong></p><ul><li><code>Sample_NoCounts.csv</code> : Sum of counts in each column.</li><li><code>Feature_Means.csv</code> : Mean in each row.</li><li><code>Feature_LogMeans.csv</code> : Log10(Mean+1) in each row.</li><li><code>Feature_SqrtMeans.csv</code> : sqrt(Mean+1) in each row.</li><li><code>Feature_Vars.csv</code> : Sample variance in each row.</li><li><code>Feature_LogVars.csv</code> : Log10(Var+1) in each row.</li><li><code>Feature_SqrtVars.csv</code> : sqrt(Var+1) in each row.</li><li><code>Feature_CV2s.csv</code> : Coefficient of Variation in each row.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/tenxsumr.jl#L1-L23">source</a></section></article><h2 id="ALGORITHM971-for-10X-HDF5"><a class="docs-heading-anchor" href="#ALGORITHM971-for-10X-HDF5">ALGORITHM971 for 10X-HDF5</a><a id="ALGORITHM971-for-10X-HDF5-1"></a><a class="docs-heading-anchor-permalink" href="#ALGORITHM971-for-10X-HDF5" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.tenxpca-Tuple{}" href="#OnlinePCA.tenxpca-Tuple{}"><code>OnlinePCA.tenxpca</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">tenxpca(; tenxfile::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;sqrt&quot;, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, colsumlist::AbstractString=&quot;&quot;, dim::Number=3, noversamples::Number=5, niter::Number=3, chunksize::Number=5000, group::AbstractString, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1.0f0)</code></pre><p>A randomized SVD.</p><p><strong>Input Arguments</strong></p><ul><li><code>tenxfile</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {sqrt,log,raw}-scaling of the value.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>noversamples</code> : The number of over-sampling.</li><li><code>niter</code> : The number of power interation.</li><li><code>chunksize</code> is the number of rows reading at once (e.g. 5000).</li><li><code>group</code> : The group name of 10XHDF5 (e.g. mm10).</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 5000) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>V</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>U</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/tenxpca.jl#L1-L33">source</a></section></article><h2 id="Sparse-Randomized-SVD-(ALGORITHM971-for-Binarized-MM-file)"><a class="docs-heading-anchor" href="#Sparse-Randomized-SVD-(ALGORITHM971-for-Binarized-MM-file)">Sparse Randomized SVD (ALGORITHM971 for Binarized MM file)</a><a id="Sparse-Randomized-SVD-(ALGORITHM971-for-Binarized-MM-file)-1"></a><a class="docs-heading-anchor-permalink" href="#Sparse-Randomized-SVD-(ALGORITHM971-for-Binarized-MM-file)" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.sparse_rsvd-Tuple{}" href="#OnlinePCA.sparse_rsvd-Tuple{}"><code>OnlinePCA.sparse_rsvd</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">sparse_rsvd(; input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;ftt&quot;, rowmeanlist::AbstractString=&quot;&quot;, rowvarlist::AbstractString=&quot;&quot;, colsumlist::AbstractString=&quot;&quot;, dim::Number=3, noversamples::Number=5, niter::Number=3, chunksize::Number=1, initW::Union{Nothing,AbstractString}=nothing, initV::Union{Nothing,AbstractString}=nothing, logdir::Union{Nothing,AbstractString}=nothing, perm::Bool=false, cper::Number=1.0f0)</code></pre><p>A randomized SVD.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.mm2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {ftt,log,raw}-scaling of the value.</li><li><code>rowmeanlist</code> : The mean of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>rowvarlist</code> : The variance of each row of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>colsumlist</code> : The sum of counts of each columns of matrix. The CSV file is generated by <code>OnlinePCA.sumr</code> functions.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>noversamples</code> : The number of over-sampling.</li><li><code>niter</code> : The number of power interation.</li><li><code>chunksize</code> is the number of rows reading at once (e.g. 1).</li><li><code>initW</code> : The CSV file saving the initial values of eigenvectors.</li><li><code>initV</code> : The CSV file saving the initial values of loadings.</li><li><code>logdir</code> : The directory where intermediate files are saved, in every evalfreq (e.g. 1) iteration.</li><li><code>perm</code> : Whether the data matrix is shuffled at random.</li><li><code>cper</code> : Count per X (e.g. CPM: Count per million &lt;1e+6&gt;)</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>V</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>U</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/sparse_rsvd.jl#L1-L32">source</a></section></article><h2 id="Exact-Out-of-Core-PCA"><a class="docs-heading-anchor" href="#Exact-Out-of-Core-PCA">Exact Out-of-Core PCA</a><a id="Exact-Out-of-Core-PCA-1"></a><a class="docs-heading-anchor-permalink" href="#Exact-Out-of-Core-PCA" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="OnlinePCA.exact_ooc_pca-Tuple{}" href="#OnlinePCA.exact_ooc_pca-Tuple{}"><code>OnlinePCA.exact_ooc_pca</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">exact_ooc_pca(; input::AbstractString=&quot;&quot;, outdir::Union{Nothing,AbstractString}=nothing, scale::AbstractString=&quot;raw&quot;, pseudocount::Number=1.0f0, dim::Number=3, chunksize::Number=1, mode::AbstractString=&quot;dense&quot;)</code></pre><p>Exact Out-of-Core PCA, which is based on normal full-rank SVD and does not assume the low-rank approximation.</p><p><strong>Input Arguments</strong></p><ul><li><code>input</code> : Julia Binary file generated by <code>OnlinePCA.csv2bin</code> or <code>OnlinPCA.mm2bin</code> function.</li><li><code>outdir</code> : The directory specified the directory you want to save the result.</li><li><code>scale</code> : {raw,log,ftt}-scaling of the value.</li><li><code>pseudocount</code> : The number specified to avoid NaN by log10(0) and used when <code>Feature_LogMeans.csv</code> &lt;log10(mean+pseudocount) value of each feature&gt; is generated.</li><li><code>dim</code> : The number of dimension of PCA.</li><li><code>chunksize</code> : The number of rows to be read at once.</li><li><code>mode</code> : &quot;dense&quot;, &quot;sparse<em>mm&quot;, or &quot;sparse</em>bincoo&quot; can be specified.</li></ul><p><strong>Output Arguments</strong></p><ul><li><code>V</code> : Eigen vectors of covariance matrix (No. columns of the data matrix × dim)</li><li><code>λ</code> : Eigen values (dim × dim)</li><li><code>U</code> : Loading vectors of covariance matrix (No. rows of the data matrix × dim)</li><li><code>Scores</code> : Principal component scores</li><li><code>ExpVar</code> : Explained variance by the eigenvectors</li><li><code>TotalVar</code> : Total variance of the data matrix</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rikenbit/OnlinePCA.jl/blob/df65589dcaaffe923906ef017565cbe35de07ddb/src/exact_ooc_pca.jl#L1-L24">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../commandline/">Command line Tool »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.9.0 on <span class="colophon-date" title="Tuesday 1 July 2025 01:39">Tuesday 1 July 2025</span>. Using Julia version 1.11.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
